### PR TITLE
add ImageDecodingNetwork for transposed image convolution

### PR DESCRIPTION
**Pull request by hnyu**
add a generic templated DecodingNetwork class for creating transposed convolution to reconstruct vector/image inputs. Could potentially be combined with VAE or Generator.

### CODE REVIEW

File: alf/utils/decoding_network.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import gin.tf
+import tensorflow as tf
+
+
+@gin.configurable
+class DecodingNetwork(tf.keras.models.Model):
```
Comment (emailweixu) [inline alf/utils/decoding_network.py]: ImageDecodingNetwork to be precise.

File: alf/utils/decoding_network.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+            output_activation_fn (tf.nn.activation): activation for the output
+                layer. Usually our image inputs are normalized to [0, 1] or [-1, 1],
+                so this function should be tf.nn.sigmoid or tf.nn.tanh.
+
+        The user is responsible for calculating the output size given
+        `start_decoding_size` and `deconv_layer_params`, and make sure that the
+        size will match the expectation. How to calculate the output size:
+            if padding=="same", then H = H1 * strides
+            if padding=="valid", then H = (H1-1) * strides + HF
+        where H = output size, H1 = input size, HF = height of kernel
```
Comment (emailweixu) [inline alf/utils/decoding_network.py]: Put description before "Args:"

File: alf/utils/decoding_network.py (hunk 1 above)
Comment (hnyu) [reply alf/utils/decoding_network.py]: This can be a decoding network for floating vector if conv_layer_params is None.

Comment (hnyu) [reply alf/utils/decoding_network.py]: renamed.

File: alf/utils/decoding_network.py (hunk 2 above)
Comment (hnyu) [reply alf/utils/decoding_network.py]: Done.

**Pull request merged**